### PR TITLE
DBHashSet Implementation

### PR DESCRIPTION
**Pull request by 0biha**
This pull request includes a DBHashSet class, whose implementation is based on the already existing DBArrayList class (the list has been replaced by a set). Furthermore, it adds a constructor to each of the existing DBCollections that supports the creation of a collection from an existing collection (by copying the elements into the wrapped list/map/set).

### CODE REVIEW

File: src/org/zoodb/api/DBArrayList.java [hunk 1]
```diff
@@ -36,8 +36,12 @@ public DBArrayList() {
 		v = new ArrayList<E>();
 	}
 	
-	public DBArrayList(int size) {
-		v = new ArrayList<E>(size);
+	public DBArrayList(int initialCapacity) {
```
Comment (0biha) [inline src/org/zoodb/api/DBArrayList.java]: I renamed this parameter because it does not affect the size() of the list.

**Comment by tzaeschke** [comment]
Thanks for your contribution.
However, if you search for example for DBArrayList in the project, there are over 100 references, so  there are some things missing:

- A test for correct java.util.Set functionality of DBHashSet
- A test for correct persistence behaviour of DBHashSet
- Tests for the new constructors
- Integration with the various serialization/deserialization classes
- Integration with the ObjectGraphTraverser
- Integration of predefined schema

Some of the tests may seem trivial, but they can detect typos and similar problems. Also, the idea is to improve performance for these special classes by special treatment, so they may always be backed by a simple ArrayList/HashMap/HashSet.

**Comment by 0biha** [comment]
Thank you for your response. I tried to add the things you mentioned. I hope my implementations make sense (more or less).

File: src/org/zoodb/internal/DataDeSerializer.java [hunk 2]
```diff
@@ -899,6 +908,18 @@ private final void deserializeDBList(List<Object> c) {
...
     
+    private final void deserializeDBSet(Set<Object> c) {
+        final int size = in.readInt();
+        c.clear();
+        if (c instanceof DBHashSet) {
+        	((DBHashSet<Object>)c).resize(size);
+        }
+        Object[] values = new Object[size];
+        for (int i=0; i < size; i++)
+            values[i] = deserializeObject();
```
Comment (tzaeschke) [inline src/org/zoodb/internal/DataDeSerializer.java]: Could you add { and } to the for-loop? THe ZooDB standard is to use braces on single-line blocks.

**Comment by tzaeschke** [comment]
Thanks, looks good (except one minor comment, see above).

I have two questions now:
1) Could you reassign the pull request to the branch called 'DBHashSet'?
2) We are thinking abot changing the license of ZooDB soon, probably to one of Apache License 2.0, LGPL or potentially AGPL. Would you agree with this change?

**Pull request closed without merging**